resallocserver/logic.py

```diff
@@ -41,34 +66,55 @@ def on(self):
         return self.query.filter(models.Resource.state != RState.ENDED)
 
     def ready(self):
-        return (self.up()
-                    .outerjoin(models.Ticket)
-                    .filter(models.Ticket.id == None))
+        """
+        Get ready resources, those which were never assigned or are released.
+        The sandbox-assigned resources are sorted above others - so they can be
+        re-used first.
+        """
+        return (self.up().filter(models.Resource.ticket_id.is_(None))
+                         .order_by(models.Resource.sandbox.is_(None)))
 
     def taken(self):
+        """
+        Get the list of all _taken_ resources == those that have OPEN ticket
+        assigned.
+        """
         return (self.up()
-                    .outerjoin(models.Ticket)
-                    .filter(models.Ticket.id != None))
+                    .filter(models.Resource.ticket_id.isnot(None)))
 
     def starting(self):
         return self.query.filter_by(state=RState.STARTING)
 
     def stats(self):
         items = {}
-        items['on']     = self.on().outerjoin(models.Ticket).all()
+        items['on']     = self.on().all()
         items['up']     = [x for x in items['on'] if x.state  == RState.UP]
-        items['ready']  = [x for x in items['up'] if x.ticket == None]
-        items['taken']  = [x for x in items['up'] if x.ticket != None]
+        items['ready']  = [x for x in items['up'] if x.ticket_id is None]
+        items['taken']  = [x for x in items['up'] if x.ticket_id is not None]
```

Comment:
```suggestion
        items['taken']  = [x for x in items['up'] if x.ticket_id]
```

Reply:
I'm not sure it is safe to expect ticket != 0 .. please keep 'is None' here to only calculate NULL values

Reply:
The original comprehensive list means that all items['taken'] values will be all except None. Everythong else is count.

Reply:
The semantics is `what resource is without assigned ticket` resp. `with assigned ticket`.  The ticket_id can probably be zero in rare cases.  I mean this:
```
In [1]: [x for x in [None, 0, "", 1] if x]                                                                                                                                                                                                   
Out[1]: [1]

In [2]: [x for x in [None, 0, "", 1] if x is not None]                                                                                                                                                                                       
Out[2]: [0, '', 1]
```

Reply:
I got it. Make sense.